bokeh/embed/util.py

```diff
@@ -49,6 +50,149 @@
+
+            If a Theme instance, use that theme, restoring any previous theme afterwards
+
+        always_new (bool, optional) :
+            Always return a new document, even in cases where it is otherwise possible
+            to use an existing document on models.
+
+    Yields:
+        Document
+
+    '''
+    # rationalize non-sequence inputs
+    if isinstance(objs, (Document, Model)):
+        objs = [objs]
+
+    # ensure inputs make sense
+    if not isinstance(objs, Sequence):
+        raise ValueError("OutputDocumentFor expects a Model, a Document, or a sequence of Models")
+
+    if len(objs) == 0:
+        raise ValueError("OutputDocumentFor expects a Model, a Document, or a sequence of Models")
+
+    if not (all(isinstance(x, Model) for x in objs) or (len(objs)==1 and isinstance(objs[0], Document))):
+        raise ValueError("OutputDocumentFor expects a Model, a Document, or a sequence of Models")
+
+    def finish(): pass
+
+    if len(objs) == 1 and isinstance(objs[0], Document):
+        doc = objs[0]
+
+    else:
+        docs = set(x.document for x in objs)
+
+        # handle a single shared document, or missing document
+        if len(docs) == 1:
+            doc = docs.pop()
+
+            # if there is no document, make one to use
+            if doc is None:
+                doc = Document()
+                for model in objs:
+                    doc.add_root(model)
+
+            # we are not using all the roots, make a quick clone for outputting purposes
+            elif set(objs) != set(doc.roots) or always_new:
+                def finish(): # NOQA
+                    _dispose_temp_doc(objs)
+                doc = _create_temp_doc(objs)
+
+            # we are using all the roots of a single doc, just use doc as-is
+            pass
+
+        # models have mixed docs, just make a quick clone
+        else:
+            def finish(): # NOQA
+                _dispose_temp_doc(objs)
+            doc = _create_temp_doc(objs)
+
+    if settings.perform_document_validation():
+        doc.validate()
+
+    _set_temp_theme(doc, apply_theme)
+
+    yield doc
+
+    _unset_temp_theme(doc)
+
+    finish()
+
+def _set_temp_theme(doc, apply_theme):
+    doc._old_theme = doc.theme
+    if apply_theme is FromCurdoc:
+        from ..io import curdoc; curdoc
+        doc.theme = curdoc().theme
+    elif apply_theme is not None:
+        doc.theme = apply_theme
+
+def _unset_temp_theme(doc):
+    doc.theme = doc._old_theme
+    del doc._old_theme
+
+def _dispose_temp_doc(models):
+    for m in models:
+        m._temp_document = None
+        for ref in m.references():
+            ref._temp_document = None
+
+def _create_temp_doc(models):
+    doc = Document()
+    for m in models:
+        doc._all_models[m._id] = m
+        m._temp_document = doc
+        for ref in m.references():
+            doc._all_models[ref._id] = ref
+            ref._temp_document = doc
+    doc._roots = models
+    return doc
```

Comment:
These four functions could be considered "quick and dirty". If you have preferences for internal-ish APIs to implement this, please speak up @mattpap